File: libs/lib-python/tests/test_linear.py
import unittest
import time
import os
from src.linear import counting_sort, radix_sort, bucket_sort, spreadsort, burstsort, flashsort, postman_sort, bead_sort, pigeonhole_sort, bucket_sort_whole_keys

BASE_DIR = os.path.dirname(__file__)
MASSA_PATH = os.path.abspath(os.path.join(BASE_DIR, "..", "..", "..", "data", "massa.txt"))

def ler_massa(path):
    with open(path, "r") as f:
        lines = f.readlines()
    
    massa = []
    tipo = None
    for line in lines:
        line = line.strip()
        if not line:
            continue
        if line.startswith("#"):
            tipo = line[1:]  # 'int' ou 'float'
            continue
        
        arr = [float(x) if '.' in x else int(x) for x in line.split()]
        massa.append((tipo, arr))
    return massa

class TestLinearSorts(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        cls.massa = ler_massa(MASSA_PATH)

    def test_algoritmos_por_algoritmo(self):
        resultados = {
            "Counting Sort": [],
            "Radix Sort": [],
            "Bucket Sort": [],
            "Spread Sort": [],
            "Burst Sort": [],
            "Flash Sort": [],
            "Postman Sort": [],
            "Bead Sort": [],
            "Pigeonhole Sort": [],
            "Bucket Sort Whole Keys": []
        }

        for idx, (tipo, arr) in enumerate(self.massa, start=1):
            if tipo == "int":
                # Counting Sort
                inicio = time.time()
                result = counting_sort(arr.copy(), max(arr) if arr else 0)
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Counting Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Radix Sort
                inicio = time.time()
                result = radix_sort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Radix Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Spread Sort
                inicio = time.time()
                result = spreadsort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Spread Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Burst Sort (converte para strings)
                arr_str = [str(x) for x in arr]
                inicio = time.time()
                result = burstsort(arr_str.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr_str)
                resultados["Burst Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Flash Sort
                inicio = time.time()
                result = flashsort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Flash Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Postman Sort
                inicio = time.time()
                result = postman_sort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Postman Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Bead Sort (apenas inteiros não-negativos)
                if all(x >= 0 for x in arr):
                    inicio = time.time()
                    result = bead_sort(arr.copy())
                    duracao = time.time() - inicio
                    correto = result == sorted(arr)
                    resultados["Bead Sort"].append({
                        "massa": idx,
                        "correto": correto,
                        "tamanho": len(arr),
                        "tempo": duracao
                    })

                # Pigeonhole Sort
                inicio = time.time()
                result = pigeonhole_sort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Pigeonhole Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Bucket Sort Whole Keys
                inicio = time.time()
                result = bucket_sort_whole_keys(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Bucket Sort Whole Keys"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

            elif tipo == "float":
                # Bucket Sort
                inicio = time.time()
                result = bucket_sort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Bucket Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Spread Sort
                inicio = time.time()
                result = spreadsort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Spread Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Burst Sort (converte para strings)
                arr_str = [str(x) for x in arr]
                inicio = time.time()
                result = burstsort(arr_str.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr_str)
                resultados["Burst Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Flash Sort
                inicio = time.time()
                result = flashsort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Flash Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Postman Sort
                inicio = time.time()
                result = postman_sort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Postman Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Pigeonhole Sort
                inicio = time.time()
                result = pigeonhole_sort(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Pigeonhole Sort"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

                # Bucket Sort Whole Keys
                inicio = time.time()
                result = bucket_sort_whole_keys(arr.copy())
                duracao = time.time() - inicio
                correto = result == sorted(arr)
                resultados["Bucket Sort Whole Keys"].append({
                    "massa": idx,
                    "correto": correto,
                    "tamanho": len(arr),
                    "tempo": duracao
                })

        # Exibe resultados organizados por algoritmo
        for alg, res_list in resultados.items():
            print(f"\n=== {alg} ===")
            for res in res_list:
                status = "OK" if res["correto"] else "FALHA"
                print(f"Massa {res['massa']}: {status}, tamanho={res['tamanho']}, tempo={res['tempo']:.6f}s")


if __name__ == "__main__":
    unittest.main()

File: libs/lib-cpp/tests/test_linear.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <algorithm>
#include <chrono>
#include <cmath>

using namespace std;

void countingSort(vector<int>& arr, int k);
void radixSort(vector<int>& arr);
void bucketSort(vector<float>& arr);
void spreadsort(vector<int>& arr, int left, int right);
void burstsort(vector<string>& arr);
void flashsort(vector<int>& arr);
void postmanSort(vector<int>& arr, int base);
void bead_sort(vector<int>& arr);
void pigeonholeSort(vector<int>& arr);
void bucketSortWholeKeys(vector<int>& arr, int b);

vector<string> split(const string& line) {
    stringstream ss(line);
    string item;
    vector<string> tokens;
    while (ss >> item)
        tokens.push_back(item);
    return tokens;
}

template<typename T>
bool verificarCorretude(const vector<T>& arr, const vector<T>& esperado) {
    return arr == esperado;
}

int main() {
    string massaPath = "../../data/massa.txt";
    ifstream file(massaPath);
    if (!file.is_open()) {
        cerr << "Erro ao abrir o arquivo: " << massaPath << endl;
        return 1;
    }

    vector<pair<string, vector<string>>> linhas;
    string tipo;
    string line;
    while (getline(file, line)) {
        if (line.empty()) continue;
        if (line[0] == '#') {
            tipo = line.substr(1); // 'int' ou 'float'
            continue;
        }
        linhas.emplace_back(tipo, split(line));
    }

    vector<string> algoritmos = {"Counting Sort", "Radix Sort", "Bucket Sort", "Spread Sort", "Burst Sort", "Flash Sort", "Postman Sort", "Bead Sort", "Pigeonhole Sort", "Bucket Sort Whole Keys"};

    for (const string& alg : algoritmos) {
        cout << "\n=== " << alg << " ===\n";

        int idx = 1;
        for (auto& [tipoLinha, valores] : linhas) {
            if ((alg == "Counting Sort" || alg == "Radix Sort" || alg == "Spread Sort" || alg == "Flash Sort" || alg == "Postman Sort" || alg == "Bead Sort" || alg == "Pigeonhole Sort" || alg == "Bucket Sort Whole Keys") && tipoLinha != "int") {
                idx++;
                continue;
            }
            if (alg == "Bucket Sort" && tipoLinha != "float") {
                idx++;
                continue;
            }
            if (alg == "Burst Sort" && tipoLinha != "int" && tipoLinha != "float") {
                idx++;
                continue;
            }

            if (alg == "Burst Sort") {
                // Burst Sort (converte int ou float para strings)
                vector<string> arr_str;
                if (tipoLinha == "int") {
                    vector<int> arr_int;
                    for (auto& v : valores) arr_int.push_back(stoi(v));
                    for (auto& num : arr_int) arr_str.push_back(to_string(num));
                } else if (tipoLinha == "float") {
                    vector<float> arr_float;
                    for (auto& v : valores) arr_float.push_back(stof(v));
                    for (auto& num : arr_float) arr_str.push_back(to_string(num));
                }

                auto esperado = arr_str;
                sort(esperado.begin(), esperado.end());

                auto start = chrono::high_resolution_clock::now();
                burstsort(arr_str);
                auto end = chrono::high_resolution_clock::now();
                chrono::duration<double> dur = end - start;

                bool correto = verificarCorretude(arr_str, esperado);
                cout << "Massa " << idx << ": " << (correto ? "OK" : "FALHA")
                     << ", tamanho=" << arr_str.size()
                     << ", tempo=" << dur.count() << "s\n";
            } else if (tipoLinha == "int") {
                vector<int> arr;
                for (auto& v : valores) arr.push_back(stoi(v));

                // Bead Sort funciona apenas com inteiros não-negativos
                if (alg == "Bead Sort") {
                    bool all_non_negative = true;
                    for (int num : arr) {
                        if (num < 0) {
                            all_non_negative = false;
                            break;
                        }
                    }
                    if (!all_non_negative) {
                        idx++;
                        continue; // Pula arrays com valores negativos
                    }
                }

                auto esperado = arr;
                sort(esperado.begin(), esperado.end());

                auto start = chrono::high_resolution_clock::now();
                if (alg == "Counting Sort") countingSort(arr, arr.empty() ? 0 : *max_element(arr.begin(), arr.end()));
                if (alg == "Radix Sort") radixSort(arr);
                if (alg == "Spread Sort") spreadsort(arr, 0, arr.size());
                if (alg == "Flash Sort") flashsort(arr);
                if (alg == "Postman Sort") postmanSort(arr, 10);
                if (alg == "Bead Sort") bead_sort(arr);
                if (alg == "Pigeonhole Sort") pigeonholeSort(arr);
                if (alg == "Bucket Sort Whole Keys") bucketSortWholeKeys(arr, 10);
                auto end = chrono::high_resolution_clock::now();
                chrono::duration<double> dur = end - start;

                bool correto = verificarCorretude(arr, esperado);
                cout << "Massa " << idx << ": " << (correto ? "OK" : "FALHA")
                     << ", tamanho=" << arr.size()
                     << ", tempo=" << dur.count() << "s\n";
            } else if (tipoLinha == "float") {
                vector<float> arr;
                for (auto& v : valores) arr.push_back(stof(v));

                auto esperado = arr;
                sort(esperado.begin(), esperado.end());

                auto start = chrono::high_resolution_clock::now();
                bucketSort(arr);
                auto end = chrono::high_resolution_clock::now();
                chrono::duration<double> dur = end - start;

                bool correto = verificarCorretude(arr, esperado);
                cout << "Massa " << idx << ": " << (correto ? "OK" : "FALHA")
                     << ", tamanho=" << arr.size()
                     << ", tempo=" << dur.count() << "s\n";
            }

            idx++;
        }
    }

    return 0;
}
